Materialized View Regression for Kafka input tables (23.12.1 version)

**Describe what's wrong**

We are using Kafka Engine table to ingest the data from Kafka to Clickhouse. Then we connect materialized view to it and store the results to target table. As we want to achieve first write wins strategy we need to check the target contents first to deduplicate incoming records. The fastest way based on benchmark was to use IN operator (JOINS were slow). To be able to do that we need to read records from source table (insert batch) twice inside materialized view. AFAIK it should be supported.

This stops to work in 23.12 with this error:

```
com.clickhouse.client.ClickHouseException: Code: 620. DB::Exception: Direct select is not allowed. To enable use setting 
`stream_like_engine_allow_direct_select`: While processing id IN ((SELECT id FROM kafka_input GROUP BY id) AS _subquery72): 
While processing id NOT IN ((SELECT id FROM deduplicate WHERE id IN (SELECT id FROM kafka_input GROUP BY id)) AS 
_subquery71). (QUERY_NOT_ALLOWED) (version 23.12.1.1368 (official build))
```

The view looks like this. Notice that we query FROM kafka_input twice (source table for materialized view)
```
CREATE MATERIALIZED VIEW IF NOT EXISTS deduplicate_mv TO deduplicate 
AS SELECT 
	id,time,any(value) AS value
FROM kafka_input 
WHERE id NOT IN (
	SELECT id FROM deduplicate WHERE id IN (
		SELECT id FROM kafka_input GROUP BY id)
	)
GROUP BY id,time;
```

Similar example using NULL engine works as expected. Here is a fiddle:
https://fiddle.clickhouse.com/a0207085-22c8-44f4-9c7f-acf751058644

**Does it reproduce on recent release?**

Yes. It is a regression of 23.12 release.


**How to reproduce**

* Use ClickHouse 23.12
* Use Kafka engine input table
* Use Materialized view that uses Kafka input table twice as in our example

**Expected behavior**

It should be possible to create such a materialized view as query is not querying kafka table itself, but the batch of inserts produced by Kafka engine table.

@filimonov here is a stack, but it just shows an exception (error) we receive over the wire.

```
10:16:00      Caused by: com.clickhouse.client.ClickHouseException: Code: 620. DB::Exception: Direct select is not allowed. To enable use setting 
`stream_like_engine_allow_direct_select`: While processing id IN ((SELECT id FROM kafka_input GROUP BY id) AS _subquery72): 
While processing id NOT IN ((SELECT id FROM deduplicate WHERE id IN (SELECT id FROM kafka_input GROUP BY id)) AS 
_subquery71). (QUERY_NOT_ALLOWED) (version 23.12.1.1368 (official build))
10:16:00      , server ClickHouseNode [uri=http://localhost:49155/metering]@-1768083100
10:16:00      	at com.clickhouse.client.ClickHouseException.of(ClickHouseException.java:168)
10:16:00      	at com.clickhouse.client.AbstractClient.execute(AbstractClient.java:282)
10:16:00      	at com.clickhouse.client.ClickHouseClientBuilder$Agent.sendOnce(ClickHouseClientBuilder.java:282)
10:16:00      	at com.clickhouse.client.ClickHouseClientBuilder$Agent.send(ClickHouseClientBuilder.java:294)
10:16:00      	at com.clickhouse.client.ClickHouseClientBuilder$Agent.execute(ClickHouseClientBuilder.java:349)
10:16:00      	at com.clickhouse.client.ClickHouseClient.executeAndWait(ClickHouseClient.java:1056)
10:16:00      	at com.clickhouse.client.ClickHouseRequest.executeAndWait(ClickHouseRequest.java:2154)
10:16:00      	at com.clickhouse.client.ClickHouseClient.lambda$send$11(ClickHouseClient.java:723)
10:16:00      	at com.clickhouse.client.ClickHouseClient.run(ClickHouseClient.java:232)
10:16:00      	at com.clickhouse.client.ClickHouseClient.lambda$submit$4(ClickHouseClient.java:284)
10:16:00      	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
10:16:00      	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
10:16:00      	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
10:16:00      	at java.base/java.lang.Thread.run(Thread.java:833)
10:16:00      Caused by: java.io.IOException: Code: 620. DB::Exception: Direct select is not allowed. To enable use setting 
`stream_like_engine_allow_direct_select`: While processing id IN ((SELECT id FROM kafka_input GROUP BY id) AS _subquery72): 
While processing id NOT IN ((SELECT id FROM deduplicate WHERE id IN (SELECT id FROM kafka_input GROUP BY id)) AS 
_subquery71). (QUERY_NOT_ALLOWED) (version 23.12.1.1368 (official build))
10:16:00  
10:16:00      	at com.clickhouse.client.http.HttpUrlConnectionImpl.checkResponse(HttpUrlConnectionImpl.java:184)
10:16:00      	at com.clickhouse.client.http.HttpUrlConnectionImpl.post(HttpUrlConnectionImpl.java:227)
10:16:00      	at com.clickhouse.client.http.ClickHouseHttpClient.send(ClickHouseHttpClient.java:124)
10:16:00      	at com.clickhouse.client.AbstractClient.execute(AbstractClient.java:280)
10:16:00      	... 12 common frames omitted
```

--- a/src/Storages/Kafka/StorageKafka.cpp
+++ b/src/Storages/Kafka/StorageKafka.cpp
@@ -8,8 +8,10 @@
 #include <DataTypes/DataTypeNullable.h>
 #include <DataTypes/DataTypeString.h>
 #include <DataTypes/DataTypesNumber.h>
+#include <Formats/FormatFactory.h>
 #include <Interpreters/Context.h>
 #include <Interpreters/InterpreterInsertQuery.h>
+#include <Interpreters/InterpreterSelectQuery.h>
 #include <Interpreters/evaluateConstantExpression.h>
 #include <Parsers/ASTCreateQuery.h>
 #include <Parsers/ASTExpressionList.h>
@@ -17,18 +19,19 @@
 #include <Parsers/ASTInsertQuery.h>
 #include <Parsers/ASTLiteral.h>
 #include <Processors/Executors/CompletedPipelineExecutor.h>
-#include <QueryPipeline/QueryPipeline.h>
+#include <Processors/QueryPlan/ISourceStep.h>
+#include <Processors/QueryPlan/QueryPlan.h>
 #include <QueryPipeline/Pipe.h>
-#include <Storages/MessageQueueSink.h>
+#include <QueryPipeline/QueryPipeline.h>
+#include <QueryPipeline/QueryPipelineBuilder.h>
 #include <Storages/Kafka/KafkaProducer.h>
 #include <Storages/Kafka/KafkaSettings.h>
 #include <Storages/Kafka/KafkaSource.h>
+#include <Storages/MessageQueueSink.h>
+#include <Storages/NamedCollectionsHelpers.h>
 #include <Storages/StorageFactory.h>
 #include <Storages/StorageMaterializedView.h>
-#include <Storages/NamedCollectionsHelpers.h>
 #include <base/getFQDNOrHostName.h>
-#include <Common/Stopwatch.h>
-#include <Common/logger_useful.h>
 #include <boost/algorithm/string/replace.hpp>
 #include <boost/algorithm/string/split.hpp>
 #include <boost/algorithm/string/trim.hpp>
@@ -37,11 +40,12 @@
 #include <Poco/Util/AbstractConfiguration.h>
 #include <Common/Exception.h>
 #include <Common/Macros.h>
+#include <Common/Stopwatch.h>
 #include <Common/formatReadable.h>
 #include <Common/getNumberOfPhysicalCPUCores.h>
+#include <Common/logger_useful.h>
 #include <Common/quoteString.h>
 #include <Common/setThreadName.h>
-#include <Formats/FormatFactory.h>
 
 #include <Storages/ColumnDefault.h>
 #include <Common/config_version.h>
@@ -172,6 +176,92 @@ struct StorageKafkaInterceptors
 
         return status;
     }
+};
+
+class ReadFromStorageKafkaStep final : public ISourceStep
+{
+public:
+    ReadFromStorageKafkaStep(
+        const Names & column_names_,
+        StoragePtr storage_,
+        const StorageSnapshotPtr & storage_snapshot_,
+        SelectQueryInfo & query_info,
+        ContextPtr context_)
+        : ISourceStep{DataStream{.header = storage_snapshot_->getSampleBlockForColumns(column_names_)}}
+        , column_names{column_names_}
+        , storage{storage_}
+        , storage_snapshot{storage_snapshot_}
+        , storage_limits{query_info.storage_limits}
+        , context{context_}
+    {
+    }
+
+    String getName() const override { return "ReadFromStorageKafka"; }
+
+    void initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings &) override
+    {
+        auto pipe = makePipe();
+
+        /// Add storage limits.
+        for (const auto & processor : pipe.getProcessors())
+            processor->setStorageLimits(storage_limits);
+
+        /// Add to processors to get processor info through explain pipeline statement.
+        for (const auto & processor : pipe.getProcessors())
+            processors.emplace_back(processor);
+
+        pipeline.init(std::move(pipe));
+    }
+
+private:
+    Pipe makePipe()
+    {
+        auto & kafka_storage = storage->as<StorageKafka &>();
+        if (kafka_storage.shutdown_called)
+            throw Exception(ErrorCodes::ABORTED, "Table is detached");
+
+        if (!context->getSettingsRef().stream_like_engine_allow_direct_select)
+            throw Exception(
+                ErrorCodes::QUERY_NOT_ALLOWED,
+                "Direct select is not allowed. To enable use setting `stream_like_engine_allow_direct_select`");
+
+        if (kafka_storage.mv_attached)
+            throw Exception(ErrorCodes::QUERY_NOT_ALLOWED, "Cannot read from StorageKafka with attached materialized views");
+
+        ProfileEvents::increment(ProfileEvents::KafkaDirectReads);
+
+        /// Always use all consumers at once, otherwise SELECT may not read messages from all partitions.
+        Pipes pipes;
+        pipes.reserve(kafka_storage.num_consumers);
+        auto modified_context = Context::createCopy(context);
+        modified_context->applySettingsChanges(kafka_storage.settings_adjustments);
+
+        // Claim as many consumers as requested, but don't block
+        for (size_t i = 0; i < kafka_storage.num_consumers; ++i)
+        {
+            /// Use block size of 1, otherwise LIMIT won't work properly as it will buffer excess messages in the last block
+            /// TODO: probably that leads to awful performance.
+            /// FIXME: seems that doesn't help with extra reading and committing unprocessed messages.
+            pipes.emplace_back(std::make_shared<KafkaSource>(
+                kafka_storage,
+                storage_snapshot,
+                modified_context,
+                column_names,
+                kafka_storage.log,
+                1,
+                kafka_storage.kafka_settings->kafka_commit_on_select));
+        }
+
+        LOG_DEBUG(kafka_storage.log, "Starting reading {} streams", pipes.size());
+        return Pipe::unitePipes(std::move(pipes));
+    }
+    ActionsDAGPtr buildFilterDAG();
+
+    const Names column_names;
+    StoragePtr storage;
+    StorageSnapshotPtr storage_snapshot;
+    std::shared_ptr<const StorageLimitsList> storage_limits;
+    ContextPtr context;
 };
 
 namespace
@@ -347,45 +437,18 @@ String StorageKafka::getDefaultClientId(const StorageID & table_id_)
     return fmt::format("{}-{}-{}-{}", VERSION_NAME, getFQDNOrHostName(), table_id_.database_name, table_id_.table_name);
 }
 
-
-Pipe StorageKafka::read(
+void StorageKafka::read(
+    QueryPlan & query_plan,
     const Names & column_names,
     const StorageSnapshotPtr & storage_snapshot,
-    SelectQueryInfo & /* query_info */,
-    ContextPtr local_context,
+    SelectQueryInfo & query_info,
+    ContextPtr query_context,
     QueryProcessingStage::Enum /* processed_stage */,
     size_t /* max_block_size */,
     size_t /* num_streams */)
 {
-    if (shutdown_called)
-        throw Exception(ErrorCodes::ABORTED, "Table is detached");
-
-    if (!local_context->getSettingsRef().stream_like_engine_allow_direct_select)
-        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED,
-                        "Direct select is not allowed. To enable use setting `stream_like_engine_allow_direct_select`");
-
-    if (mv_attached)
-        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED, "Cannot read from StorageKafka with attached materialized views");
-
-    ProfileEvents::increment(ProfileEvents::KafkaDirectReads);
-
-    /// Always use all consumers at once, otherwise SELECT may not read messages from all partitions.
-    Pipes pipes;
-    pipes.reserve(num_consumers);
-    auto modified_context = Context::createCopy(local_context);
-    modified_context->applySettingsChanges(settings_adjustments);
-
-    // Claim as many consumers as requested, but don't block
-    for (size_t i = 0; i < num_consumers; ++i)
-    {
-        /// Use block size of 1, otherwise LIMIT won't work properly as it will buffer excess messages in the last block
-        /// TODO: probably that leads to awful performance.
-        /// FIXME: seems that doesn't help with extra reading and committing unprocessed messages.
-        pipes.emplace_back(std::make_shared<KafkaSource>(*this, storage_snapshot, modified_context, column_names, log, 1, kafka_settings->kafka_commit_on_select));
-    }
-
-    LOG_DEBUG(log, "Starting reading {} streams", pipes.size());
-    return Pipe::unitePipes(std::move(pipes));
+    query_plan.addStep(std::make_unique<ReadFromStorageKafkaStep>(
+        column_names, shared_from_this(), storage_snapshot, query_info, std::move(query_context)));
 }
 
 

--- a/src/Storages/Kafka/StorageKafka.h
+++ b/src/Storages/Kafka/StorageKafka.h
@@ -20,6 +20,7 @@ namespace DB
 {
 
 class StorageSystemKafkaConsumers;
+class ReadFromStorageKafkaStep;
 
 struct StorageKafkaInterceptors;
 
@@ -48,7 +49,8 @@ public:
     void startup() override;
     void shutdown(bool is_drop) override;
 
-    Pipe read(
+    void read(
+        QueryPlan & query_plan,
         const Names & column_names,
         const StorageSnapshotPtr & storage_snapshot,
         SelectQueryInfo & query_info,
@@ -86,6 +88,8 @@ public:
     SafeConsumers getSafeConsumers() { return {shared_from_this(), std::unique_lock(mutex), consumers};  }
 
 private:
+    friend class ReadFromStorageKafkaStep;
+
     // Configuration and state
     std::unique_ptr<KafkaSettings> kafka_settings;
     Macros::MacroExpansionInfo macros_info;

--- a/tests/integration/test_storage_kafka/test.py
+++ b/tests/integration/test_storage_kafka/test.py
@@ -4930,6 +4930,77 @@ def test_formats_errors(kafka_cluster):
         instance.query(f"DROP TABLE test.{table_name}")
         instance.query("DROP TABLE test.view")
 
+def test_multiple_read_in_materialized_views(kafka_cluster, max_retries=15):
+    admin_client = KafkaAdminClient(
+        bootstrap_servers="localhost:{}".format(kafka_cluster.kafka_port)
+    )
+
+    topic = "multiple_read_from_mv"
+    kafka_create_topic(admin_client, topic)
+
+    instance.query(
+        f"""
+        DROP TABLE IF EXISTS test.kafka_multiple_read_input;
+        DROP TABLE IF EXISTS test.kafka_multiple_read_table;
+        DROP TABLE IF EXISTS test.kafka_multiple_read_mv;
+
+        CREATE TABLE test.kafka_multiple_read_input (id Int64)
+        ENGINE = Kafka
+        SETTINGS
+            kafka_broker_list = 'kafka1:19092',
+            kafka_topic_list = '{topic}',
+            kafka_group_name = '{topic}',
+            kafka_format = 'JSONEachRow';
+
+        CREATE TABLE test.kafka_multiple_read_table (id Int64)
+        ENGINE = MergeTree
+        ORDER BY id;
+
+
+        CREATE MATERIALIZED VIEW IF NOT EXISTS test.kafka_multiple_read_mv TO test.kafka_multiple_read_table AS
+        SELECT id
+        FROM test.kafka_multiple_read_input
+        WHERE id NOT IN (
+            SELECT id
+            FROM test.kafka_multiple_read_table
+            WHERE id IN (
+                SELECT id
+                FROM test.kafka_multiple_read_input
+            )
+        );
+        """
+    )
+
+    kafka_produce(kafka_cluster, topic, [json.dumps({"id": 42}), json.dumps({"id": 43})])
+
+    expected_result = "42\n43\n"
+    res = instance.query_with_retry(
+        f"SELECT id FROM test.kafka_multiple_read_table ORDER BY id",
+        retry_count=30,
+        sleep_time=0.5,
+        check_callback=lambda res: res == expected_result,
+    )
+    assert res == expected_result
+
+    # Verify that the query deduplicates the records as it meant to be
+    messages = []
+    for i in range(0,10):
+        messages.append(json.dumps({"id": 42}))
+        messages.append(json.dumps({"id": 43}))
+
+    messages.append(json.dumps({"id": 44}))
+
+    kafka_produce(kafka_cluster, topic, messages)
+
+    expected_result = "42\n43\n44\n"
+    res = instance.query_with_retry(
+        f"SELECT id FROM test.kafka_multiple_read_table ORDER BY id",
+        retry_count=30,
+        sleep_time=0.5,
+        check_callback=lambda res: res == expected_result,
+    )
+    assert res == expected_result
+
 
 if __name__ == "__main__":
     cluster.start()

--- a/tests/integration/test_storage_kafka/test.py
+++ b/tests/integration/test_storage_kafka/test.py
@@ -892,12 +892,14 @@ def test_kafka_formats(kafka_cluster):
 """
 
     expected_rows_count = raw_expected.count("\n")
-    instance.query_with_retry(
+    result_checker = lambda res: res.count("\n") == expected_rows_count
+    res = instance.query_with_retry(
         f"SELECT * FROM test.kafka_{list(all_formats.keys())[-1]}_mv;",
         retry_count=30,
         sleep_time=1,
-        check_callback=lambda res: res.count("\n") == expected_rows_count,
+        check_callback=result_checker,
     )
+    assert result_checker(res)
 
     for format_name, format_opts in list(all_formats.items()):
         logging.debug(("Checking {}".format(format_name)))
@@ -3808,12 +3810,14 @@ def test_kafka_formats_with_broken_message(kafka_cluster):
 """
 
     expected_rows_count = raw_expected.count("\n")
-    instance.query_with_retry(
+    result_checker = lambda res: res.count("\n") == expected_rows_count
+    res = instance.query_with_retry(
         f"SELECT * FROM test.kafka_data_{list(all_formats.keys())[-1]}_mv;",
         retry_count=30,
         sleep_time=1,
-        check_callback=lambda res: res.count("\n") == expected_rows_count,
+        check_callback=result_checker,
     )
+    assert result_checker(res)
 
     for format_name, format_opts in list(all_formats.items()):
         logging.debug(f"Checking {format_name}")

--- a/tests/integration/test_storage_kafka/test.py
+++ b/tests/integration/test_storage_kafka/test.py
@@ -4934,6 +4934,7 @@ def test_formats_errors(kafka_cluster):
         instance.query(f"DROP TABLE test.{table_name}")
         instance.query("DROP TABLE test.view")
 
+
 def test_multiple_read_in_materialized_views(kafka_cluster, max_retries=15):
     admin_client = KafkaAdminClient(
         bootstrap_servers="localhost:{}".format(kafka_cluster.kafka_port)
@@ -4975,7 +4976,9 @@ def test_multiple_read_in_materialized_views(kafka_cluster, max_retries=15):
         """
     )
 
-    kafka_produce(kafka_cluster, topic, [json.dumps({"id": 42}), json.dumps({"id": 43})])
+    kafka_produce(
+        kafka_cluster, topic, [json.dumps({"id": 42}), json.dumps({"id": 43})]
+    )
 
     expected_result = "42\n43\n"
     res = instance.query_with_retry(
@@ -4988,7 +4991,7 @@ def test_multiple_read_in_materialized_views(kafka_cluster, max_retries=15):
 
     # Verify that the query deduplicates the records as it meant to be
     messages = []
-    for i in range(0,10):
+    for i in range(0, 10):
         messages.append(json.dumps({"id": 42}))
         messages.append(json.dumps({"id": 43}))
 

